# Task 2.1 — One Knob at a Time (15 min)
Pick **ONE**: `max_depth ∈ {2, 4, 8, 12}` **or** `learning_rate ∈ {0.01, 0.1, 0.3}`.
Retrain per value, fill the table, then **one sentence**: what changed + your hypothesis why.

In [2]:
import sys, pathlib
p = pathlib.Path.cwd()
while not (p / 'src').exists():
    p = p.parent
sys.path.insert(0, str(p / 'src')); REPO = p
import warnings; warnings.filterwarnings('ignore')

In [4]:
from course_utils import data
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

df = data.load_processed('telco_clean')
X, y = data.prep_features(df)
S = data.make_splits(X, y)

def run(**params):
    base = dict(n_estimators=200, max_depth=4, learning_rate=0.1,
                eval_metric='auc', random_state=42)
    base.update(params)
    m = XGBClassifier(**base).fit(S['X_train'], S['y_train'])
    return (roc_auc_score(S['y_train'], m.predict_proba(S['X_train'])[:, 1]),
            roc_auc_score(S['y_val'], m.predict_proba(S['X_val'])[:, 1]))

# example: run(max_depth=2)

In [5]:
# your experiment here — loop over YOUR chosen knob and print train/val AUC per value
for depth in [2, 4, 8, 12]:
    train_auc, val_auc = run(max_depth=depth)
    print(f"max_depth={depth}: train AUC={train_auc:.4f}, val AUC={val_auc:.4f}")

max_depth=2: train AUC=0.8714, val AUC=0.8497
max_depth=4: train AUC=0.9213, val AUC=0.8421
max_depth=8: train AUC=0.9930, val AUC=0.8330
max_depth=12: train AUC=0.9995, val AUC=0.8232


## Results
| value | train AUC | val AUC |
|---|---|---|
| 2 | 0.8714 | 0.8497 |
| 4 | 0.9213 | 0.8421 |
| 8 | 0.9930 | 0.8330 |
| 12 | 0.9995 | 0.8232 |

**What changed & why (1 sentence):**
As max_depth increased the training AUC increased while the validation AUC decreased. This indicates that the model became more complex and started to overfit